client/webserver/api.go

```diff
@@ -1616,6 +1643,179 @@ func (s *WebServer) apiSendShielded(w http.ResponseWriter, r *http.Request) {
 	writeJSON(w, simpleAck(), s.indent)
 }
 
+func (s *WebServer) apiStartMarketMaking(w http.ResponseWriter, r *http.Request) {
+	var form struct {
+		AppPW encode.PassBytes `json:"appPW"`
+	}
+	if !readPost(w, r, &form) {
+		s.writeAPIError(w, fmt.Errorf("failed to read form"))
+		return
+	}
+
+	cfg, err := s.getMarketMakingConfig()
+	if err != nil {
+		s.writeAPIError(w, fmt.Errorf("error getting market making config: %v", err))
+		return
+	}
+
+	err = s.mm.Run(s.ctx, cfg, form.AppPW)
+	if err != nil {
+		s.writeAPIError(w, fmt.Errorf("Error starting market making: %v", err))
+		return
+	}
+
+	writeJSON(w, simpleAck(), s.indent)
+}
+
+func (s *WebServer) apiStopMarketMaking(w http.ResponseWriter, r *http.Request) {
+	s.mm.Stop()
+	writeJSON(w, simpleAck(), s.indent)
+}
+
+func (s *WebServer) getMarketMakingConfig() ([]*mm.BotConfig, error) {
+	cfg := []*mm.BotConfig{}
+	if s.mmCfgPath != "" {
+		data, err := ioutil.ReadFile(s.mmCfgPath)
+		if err == nil {
+			err = json.Unmarshal(data, &cfg)
+			if err != nil {
+				return nil, fmt.Errorf("error unmarshalling market making config: %v", err)
+			}
+		} else if !os.IsNotExist(err) {
+			return nil, fmt.Errorf("error reading file: %v", err)
+		}
+	}
+
+	return cfg, nil
+}
+
+func (s *WebServer) apiMarketMakingConfig(w http.ResponseWriter, r *http.Request) {
+	cfg, err := s.getMarketMakingConfig()
+	if err != nil {
+		s.writeAPIError(w, fmt.Errorf("error getting market making config: %v", err))
+		return
+	}
+
+	writeJSON(w, &struct {
+		OK  bool            `json:"ok"`
+		Cfg []*mm.BotConfig `json:"cfg"`
+	}{
+		OK:  true,
+		Cfg: cfg,
+	}, s.indent)
+}
+
+func (s *WebServer) apiUpdateMarketMakingConfig(w http.ResponseWriter, r *http.Request) {
+	var updatedCfg *mm.BotConfig
+	if !readPost(w, r, &updatedCfg) {
+		s.writeAPIError(w, fmt.Errorf("failed to read config"))
+		return
+	}
+
+	cfg, err := s.getMarketMakingConfig()
+	if err != nil {
+		s.writeAPIError(w, fmt.Errorf("error getting market making config: %v", err))
+		return
+	}
+
+	var updated bool
+	for i, c := range cfg {
+		if c.Host == updatedCfg.Host && c.QuoteAsset == updatedCfg.QuoteAsset && c.BaseAsset == updatedCfg.BaseAsset {
+			cfg[i] = updatedCfg
+			updated = true
+			break
+		}
+	}
+	if !updated {
+		cfg = append(cfg, updatedCfg)
+	}
+
+	data, err := json.MarshalIndent(cfg, "", "    ")
+	if err != nil {
+		s.writeAPIError(w, fmt.Errorf("error marshalling market making config: %v", err))
+		return
+	}
+
+	err = ioutil.WriteFile(s.mmCfgPath, data, 0644)
+	if err != nil {
+		s.writeAPIError(w, fmt.Errorf("error writing market making config: %v", err))
+		return
+	}
```

Comment:
Looks good. I'm also adding an alternate config file path to `MarketMaker.Run` so that the RPC server can use whatever file the user specifies.

Reply:
I did it this way because when using the RPC server, I think it would be more convenient if you could just specify the path to the config file instead of having to first update the config and then run it. Should `Run` take an alternate config file path and if it is nil, then